### ISSUE TITLE
Allow narrowing properties retrieved from types with index signatures

### ISSUE THREAD

**Issue by kujon** [Microsoft/TypeScript#17567]
**TypeScript Version:**  2.4.0

**Code**

```ts
type A = {
    [index: string]: [number, number] | null;
};

const value: A = {
    foo: [42, 42]
};

const extractedValue = value.foo !== null ?
    // Object is possibly 'null'.
    value.foo[0] :
    5;
```

**Expected behavior:**
Null check is correctly performed and the compiler allows for accessing 

**Actual behavior:**
// Object is possibly 'null'.

**Comment by DanielRosenwasser** [comment]
This would require us to synthesize some sort of knowledge about a property that's never been declared on the type `A`.

**Comment by mhegazy** [comment]
Should be doable. `getFlowTypeOfReference` operates on nodes, so we should be able to pass in the node, along with the declared type from the index signature.

**Comment by sandersn** [comment]
Fix is up at #19838.
